src/main/kotlin/rx/lang/kotlin/observables.kt

```diff
@@ -58,34 +58,20 @@ public fun <T, R> Observable<T>.lift(operator : (Subscriber<in R>) -> Subscriber
 /**
  * Returns [Observable] that requires all objects to be non null. Raising [NullPointerException] in case of null object
  */
-public fun <T : Any> Observable<T?>.requireNoNulls() : Observable<T> = lift { s ->
-    subscriber<T?>().
-            onCompleted { s.onCompleted() }.
-            onError { t -> s.onError(t) }.
-            onNext { v -> if (v == null) throw NullPointerException("null element found in rx observable") else s.onNext(v) }
-}
+public fun <T : Any> Observable<T?>.requireNoNulls() : Observable<T> = map { it ?: throw NullPointerException("null element found in rx observable") }
 
 /**
  * Returns [Observable] with non-null generic type T. Returned observable filter out null values
  */
-public fun <T : Any> Observable<T?>.filterNotNull() : Observable<T> = lift { s ->
-    subscriber<T?>().
-            onCompleted { s.onCompleted() }.
-            onError { t -> s.onError(t) }.
-            onNext { v -> if (v != null) s.onNext(v) }
-}
+public fun <T : Any> Observable<T?>.filterNotNull(): Observable<T> = filter { it != null }.map { it as T }
```

Comment:
filter{it!=null} -> (we still have T?, but T in runtime because no null values will be passed) (compiler can't smart cast here, because kotlin is not aware of rx filter), so we need somehow  cast value ->
map {it as T} will give us T! , you're right, but it compiles correct, and we won't get runtime exceptions.

Some kind of hack i know. Do you have better solution?